### PR TITLE
Building new architecture sources on Windows

### PR DESCRIPTION
**Pull request by mganandraj**
On Windows there are limits on number of character in file paths and in command lines
Ref: https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#Path-Length-Limits
NDK allows circumventing command line limits using response(RSP) files as inputs using NDK_APP_SHORT_COMMANDS flag.

Windows can support long file paths if configured through registry or by prefixing all file paths with a special character sequence
The latter requires changes in NDK. And there are tools in NDK (AR) which is not able to handle long paths (>256) even after setting the registry key.
The new architecutre source tree is too deep, and the object file naming conventions in NDK makes the matters worse, by producing incredibly long file paths.
Other solutions such as symlinking source code etc. didn't work as expected, and makes the build scripts complicated and hard to manage.
This change temporarily works around the issue by placing the temporary build outputs as short a path as possible within the project path.

## Changelog

[CATEGORY] [TYPE] - Message

### CODE REVIEW

File: ReactAndroid/build.gradle [hunk 1]
```diff
@@ -318,6 +318,23 @@ android {
...
+                // NDK allows circumventing command line limits using response(RSP) files as inputs using NDK_APP_SHORT_COMMANDS flag.
+                // 
+                // Windows can support long file paths if configured through registry or by prefixing all file paths with a special character sequence 
+                // The latter requires changes in NDK. And there are tools in NDK (AR) which is not able to handle long paths (>256) even after setting the registry key.
+                // The new architecutre source tree is too deep, and the object file naming conventions in NDK makes the matters worse, by producing incredibly long file paths.
+                // Other solutions such as symlinking source code etc. didn't work as expected, and makes the build scripts complicated and hard to manage.
+                // This change temporarily works around the issue by placing the temporary build outputs as short a path as possible within the project path.
+                if (Os.isFamily(Os.FAMILY_WINDOWS) && 
+                    project.hasProperty("newArchEnabled") && 
+                    project.newArchEnabled == "true") {
```
Comment (cortinico) [inline ReactAndroid/build.gradle:333]: Those two are probably unnecessary. You will be including `ReactAndroid/build.gradle` only when using the new arch.

File: ReactAndroid/build.gradle [hunk 2]
```diff
@@ -318,6 +318,23 @@ android {
...
+                    var rootProjPath = rootProject.projectDir.getParent().toString().trim()
```
Comment (cortinico) [inline ReactAndroid/build.gradle:334]: I'm not sure what was the intention here, but you're essentially picking the root folder of a RN project. This is risky as you're placing intermediate build artifacts inside a `.cxx` hidden folder in the root of the RN project.

Potentially this should be just something like: `$buildDir` (I would not go to the root folder as this is Android specific, unless you have a strong reason for that), so you'll be placing everything inside `$buildDir\\.cxx`.

File: ReactAndroid/build.gradle [hunk 3]
```diff
@@ -318,6 +318,23 @@ android {
...
+                    arguments "NDK_OUT=$rootProjPath\\.cxx",
+                              "NDK_APP_SHORT_COMMANDS=true"
```
Comment (cortinico) [inline ReactAndroid/build.gradle:336]: nit: What would be the impact of having those for Linux and Mac?

Also you might need the same directives for the `template/android/app/build.gradle` file?

**Review by cortinico — COMMENTED** [review]
Great stuff! Thanks for looking into this @mganandraj and thanks for leaving the great explanation. Left a couple of comments that needs to be addressed before we can land this

File: ReactAndroid/build.gradle (hunk 2 above)
Comment (mganandraj) [reply ReactAndroid/build.gradle:334]: Essentially, i'm trying to place the object files in the shortest possible path within the root project .. Using $buildDir typically adds ~15 characters to the object file paths which makes them unavailable for llvm-aar ..

File: ReactAndroid/build.gradle (hunk 3 above)
Comment (mganandraj) [reply ReactAndroid/build.gradle:336]: NDK_APP_SHORT_COMMANDS should for other platforms too .. I found rsp files to be very useful for debugging and analyzing build issues, as it lists all the inputs (files, flags etc.) to various tools ..

Sure, ill update template/android/app/build.gradle too ..

File: ReactAndroid/build.gradle (hunk 1 above)
Comment (mganandraj) [reply ReactAndroid/build.gradle:333]: I think we need to keep those check for those who directly build the react-native repo to launch rntester, but don't want to enable new arch..

File: ReactAndroid/build.gradle (hunk 2 above)
Comment (mganandraj) [reply ReactAndroid/build.gradle:334]: I will need to debug llvm-ar to understand why it can't support long path .. May be it's fixed in a newer ndk ..

File: ReactAndroid/build.gradle (hunk 3 above)
Comment (cortinico) [reply ReactAndroid/build.gradle:336]: Let's keep it for Win only for now 👍 Thanks for clarifying

File: ReactAndroid/build.gradle (hunk 2 above)
Comment (cortinico) [reply ReactAndroid/build.gradle:334]: > adds ~15 characters to the object file paths which makes them unavailable for llvm-aar ..

Could you show me an example of the resulting path? I'm trying to understand where are we hitting the limit.

I'd say that definitely this has to live inside the `android` folder (not in the top level). Maybe the `android/.cxx` folder could work. Still users on Win will suffer stale builds problems if this doesn't live inside the `build/` folder.

Comment (mganandraj) [reply ReactAndroid/build.gradle:334]: This is an example path ..
D:\TestnewArc\node_modules\react-native\ReactAndroid\build\intermediates\cxx\Debug\351u4r48/obj/local/x86/objs-debug/rrc_switch/E_/gh-droid/testapp/test68/node_modules/react-native/ReactAndroid/__/ReactCommon/react/renderer/components/switch/androidswitch/react/renderer/components/androidswitch/AndroidSwitchShadowNode.o

This segment is ~230 characters;
"local/x86/objs-debug/rrc_switch/E_/gh-droid/testapp/test68/node_modules/react-native/ReactAndroid/__/ReactCommon/react/renderer/components/switch/androidswitch/react/renderer/components/androidswitch/AndroidSwitchShadowNode.o" .. and we have ~30 characters for the prefix ..

**Comment by cortinico** [comment]
Could you also point this PR to `main`? It's easier to merge it there and cherry-pick it into the release branch.

**Comment by cortinico** [comment]
> Could you also point this PR to `main`? It's easier to merge it there and cherry-pick it into the release branch.

Sorry, I've just realized this can't target `main` as we're on CMake there. I believe we should be fine with it so this patch should not even need to be backported to `main` (or maybe only the `template/android` changes).

**Comment by mganandraj** [comment]
> > Could you also point this PR to `main`? It's easier to merge it there and cherry-pick it into the release branch.
> 
> Sorry, I've just realized this can't target `main` as we're on CMake there. I believe we should be fine with it so this patch should not even need to be backported to `main` (or maybe only the `template/android` changes).

Makese sense .. Ill limit the PR to 'template/android' changes ..

**Comment by mikehardy** [comment]
Here's my full trace testing v3 of this patch from a couple hours ago (last push I think)
https://github.com/facebook/react-native/pull/33611#issuecomment-1096676299

TL;DR: it worked to a point but still may have some issues? (that is, it worked for project rooted in C:\RN068 but failed with project rooted in C:\Users\micro\work\react-random\RN068
